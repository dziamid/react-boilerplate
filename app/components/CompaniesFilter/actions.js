
export function changeValue(value) {
  return {
    type: 'COMPANIES_FILTER_VALUE_CHANGED',
    payload: value,
  };
}

export function selectCompany(company) {
  return {
    type: 'COMPANIES_FILTER_SELECT_COMPANY',
    payload: company,
  };
}

export function unselectCompany(company) {
  return {
    type: 'COMPANIES_FILTER_UNSELECT_COMPANY',
    payload: company,
  };
}
