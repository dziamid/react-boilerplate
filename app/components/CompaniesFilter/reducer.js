import { fromJS } from 'immutable';

const initialState = fromJS({
  value: 'favorite',
  companies: [],
});

function companiesFilterReducer(state = initialState, action) {
  const companies = state.get('companies');

  switch (action.type) {
    case 'COMPANIES_FILTER_VALUE_CHANGED':
      return state.set('value', action.payload);

    case 'COMPANIES_FILTER_SELECT_COMPANY':
      return state.set('companies', companies.push(action.payload));

    case 'COMPANIES_FILTER_UNSELECT_COMPANY':
      return state.set('companies', companies.delete(companies.indexOf(action.payload)));
    default:
      return state;
  }
}

export default companiesFilterReducer;
