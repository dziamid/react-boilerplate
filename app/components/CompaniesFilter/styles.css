.companies {
  margin-left: 70px;
}
