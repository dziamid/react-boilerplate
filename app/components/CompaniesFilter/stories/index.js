import React, { Component } from 'react';
import { storiesOf, action } from '@kadira/storybook';
import CompaniesFilter from '../index';

class Header extends Component {
  constructor() {
    super();
    this.handleFilterChange = this.handleFilterChange.bind(this);
    this.state = {
      value: 'custom',
    };
  }

  handleFilterChange(value) {
    this.setState({ value });
    action('onFilterChange')(value);
  }

  render() {
    const props = {
      value: this.state.value,
      onFilterChange: this.handleFilterChange,
    };

    return (
      <div>
        {React.cloneElement(this.props.children, props)}
      </div>
    );
  }
}

storiesOf('CompaniesFilter', module)
  .addDecorator((getStory) => <Header>{getStory()}</Header>)
  .add('open', () => (
    <CompaniesFilter />
  ));
