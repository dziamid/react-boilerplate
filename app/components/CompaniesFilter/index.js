import React, { PropTypes } from 'react';
import Menu from 'material-ui/Menu';
import MenuItem from 'material-ui/MenuItem';
import Divider from 'material-ui/Divider';
import Checkbox from 'material-ui/Checkbox';
import RadioButtonOn from 'material-ui/svg-icons/toggle/radio-button-checked';
import RadioButtonOff from 'material-ui/svg-icons/toggle/radio-button-unchecked';
import SettingsIcon from 'material-ui/svg-icons/action/settings';
import AddIcon from 'material-ui/svg-icons/content/add';
import styles from './styles.css';

export default class CompaniesFilter extends React.Component {

  static propTypes = {
    value: PropTypes.string,
    onFilterChange: PropTypes.func.isRequired,
    onCompanySelected: PropTypes.func.isRequired,
    onCompanyUnselected: PropTypes.func.isRequired,
    selectedCompanies: PropTypes.array,
    companies: PropTypes.array,
  };

  static defaultProps = {
    value: 'all',
    selectedCompanies: [],
    companies: [],
  };

  constructor(props) {
    super(props);
    this.handleFilterChange = this.handleFilterChange.bind(this);
    this.handleCompanyChanged = this.handleCompanyChanged.bind(this);
  }

  handleFilterChange(event, value) {
    if (value === undefined) {
      return;
    }

    this.props.onFilterChange(value);
  }

  handleCompanyChanged(company, isSelected) {
    const prop = isSelected ? this.props.onCompanySelected : this.props.onCompanyUnselected;
    prop(company);
  }

  radioIcon(value) {
    return value === this.props.value ? <RadioButtonOn /> : <RadioButtonOff />;
  }

  render() {
    const customMenuItems = this.props.value === 'custom' ? (
      <div data-id="companies" className={styles.companies}>
        { this.props.companies.map((c, i) => (
          <Checkbox
            key={i}
            label={c.name}
            value={c.id}
            onCheck={(event, isChecked) => this.handleCompanyChanged(c, isChecked)}
            checked={this.props.selectedCompanies.includes(c)}
          />
        ))}
      </div>
    ) : null;

    return (
      <Menu value={this.props.value} onChange={this.handleFilterChange}>
        <MenuItem
          onTouchTap={() => console.log('MenuItem.onTouchTap')}
          value="all"
          primaryText="Show All Companies"
          leftIcon={this.radioIcon('all')}
        />

        <Divider inset />

        <MenuItem
          value="custom"
          primaryText="Show Selected Companies"
          leftIcon={this.radioIcon('custom')}
        />
        { customMenuItems }

        <Divider inset />
        <MenuItem
          value="favorite"
          primaryText="Show Favorite Companies"
          leftIcon={this.radioIcon('favorite')}
        />
        <Divider />
        <MenuItem
          primaryText="Manage Companies"
          leftIcon={<SettingsIcon />}
        />
        <Divider />
        <MenuItem
          primaryText="Add New Company"
          leftIcon={<AddIcon />}
        />
      </Menu>

    );
  }
}
