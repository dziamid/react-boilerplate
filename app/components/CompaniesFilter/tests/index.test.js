/**
 * Testing our Button component
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import CompaniesFilter from '../index';

import getMuiTheme from 'material-ui/styles/getMuiTheme';
import { default as theme } from 'theme';

const noop = () => {
};

const mountOpts = {
  context: { muiTheme: getMuiTheme(theme) },
  childContextTypes: { muiTheme: React.PropTypes.object },
};

import injectTapEvent from 'utils/react-tap-event-plugin';
injectTapEvent();

const defaultProps = {
  value: 'favorite',
  onFilterChange: noop,
  onCompanySelected: noop,
  onCompanyUnselected: noop,
};

describe('<CompaniesFilter />', () => {
  it('should render with default props', () => {
    const wrapper = shallow(
      <CompaniesFilter {...defaultProps} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  xit('should call onFilterChange when user selects another menu item', () => {
    const onChangeSpy = jest.fn();
    const wrapper = mount(
      <CompaniesFilter {...defaultProps} onFilterChange={onChangeSpy} />,
      mountOpts,
    );

    // todo: use attribute and(or) constructor selectors,
    // for some reason '[value="all"]' selector yields no result
    const firstMenuItem = wrapper.find('MenuItem').first();
    console.log(firstMenuItem.debug());
    firstMenuItem.simulate('click');

    expect(onChangeSpy).toHaveBeenCalledWith('all');
  });

  xit('should call onCompanySelected when users checks a company in custom filter');
  xit('should call onCompanyUnselected when users unchecks a company in custom filter');
});
