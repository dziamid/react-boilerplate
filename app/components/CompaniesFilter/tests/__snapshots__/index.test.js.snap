exports[`<CompaniesFilter /> should render with default props 1`] = `
<Menu
  autoWidth={true}
  desktop={false}
  disableAutoFocus={false}
  initiallyKeyboardFocused={false}
  maxHeight={null}
  multiple={false}
  onChange={[Function]}
  onEscKeyDown={[Function]}
  onItemTouchTap={[Function]}
  onKeyDown={[Function]}
  value="favorite">
  <MenuItem
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    leftIcon={<ToggleRadioButtonUnchecked />}
    primaryText="Show All Companies"
    value="all" />
  <Divider
    inset={true} />
  <MenuItem
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    leftIcon={<ToggleRadioButtonUnchecked />}
    primaryText="Show Selected Companies"
    value="custom" />
  <Divider
    inset={true} />
  <MenuItem
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    leftIcon={<ToggleRadioButtonChecked />}
    primaryText="Show Favorite Companies"
    value="favorite" />
  <Divider
    inset={false} />
  <MenuItem
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    leftIcon={<ActionSettings />}
    primaryText="Manage Companies" />
  <Divider
    inset={false} />
  <MenuItem
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    leftIcon={<ContentAdd />}
    primaryText="Add New Company" />
</Menu>
`;
